server/src/main/java/org/opensearch/indices/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,264 @@
+        }
+    }
+
+    protected KeyStoreStats stats;
+    protected RoaringBitmap rbm;
+    private HashSet<Integer> collidedInts;
+    protected RBMSizeEstimator sizeEstimator;
+    protected final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    protected final Lock readLock = lock.readLock();
+    protected final Lock writeLock = lock.writeLock();
+
+    RBMIntKeyLookupStore(int modulo, long memSizeCapInBytes) {
+        this.modulo = modulo;
+        sizeEstimator = new RBMSizeEstimator(modulo);
+        this.stats = new KeyStoreStats(memSizeCapInBytes, calculateMaxNumEntries(memSizeCapInBytes));
+        this.rbm = new RoaringBitmap();
+        collidedInts = new HashSet<>();
+    }
+
+    protected int calculateMaxNumEntries(long memSizeCapInBytes) {
+        if (memSizeCapInBytes == 0) {
+            return Integer.MAX_VALUE;
+        }
+        return sizeEstimator.getNumEntriesFromSizeInBytes(memSizeCapInBytes);
+    }
+
+    protected final int transform(int value) {
+        return modulo == 0 ? value : value % modulo;
+    }
+
+    protected void handleCollisions(int transformedValue) {
+        stats.numCollisions.inc();
+        collidedInts.add(transformedValue);
+    }
+
+    @Override
+    public boolean add(Integer value) throws Exception {
+        if (value == null) {
+            return false;
+        }
+        writeLock.lock();
+        stats.numAddAttempts.inc();
+        try {
+            if (stats.size == stats.maxNumEntries) {
+                stats.atCapacity = true;
+                return false;
+            }
+            int transformedValue = transform(value);
+            boolean alreadyContained = contains(value);
+            if (!alreadyContained) {
+                rbm.add(transformedValue);
+                stats.size++;
+                return true;
+            }
+            handleCollisions(transformedValue);
+            return false;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public boolean contains(Integer value) throws Exception {
+        if (value == null) {
+            return false;
+        }
+        int transformedValue = transform(value);
+        readLock.lock();
+        try {
+            return rbm.contains(transformedValue);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public Integer getInternalRepresentation(Integer value) {
+        if (value == null) {
+            return 0;
+        }
+        return Integer.valueOf(transform(value));
+    }
+
+    @Override
+    public boolean remove(Integer value) throws Exception {
+        if (value == null) {
+            return false;
+        }
+        int transformedValue = transform(value);
+        readLock.lock();
+        try {
+            if (!contains(value)) {
+                return false;
+            }
+            stats.numRemovalAttempts.inc();
+            if (collidedInts.contains(transformedValue)) {
```

Comment:
I guess we will have to eventually remove values from collidedInts as well?

For example, I have two keys A and B mapping to same hash C.
Now I try removing A, but it actually doesn't remove because B also maps to C. Return false.
Now I try removing B, here it agains return false which seems wrong. As ideally it should have removed from RBM itself as now no key maps to C. 

I am guessing we might need counter as well along with HashSet to track number of collisions and keep decreasing/increasing it as we receive remove/add calls for a particular colliding value. When it reaches 1, remove the value altogether. Let me know what you think.

Reply:
Sounds good - like we talked about yesterday, I'll change this to use the combination of a counter for collided value C, as well as a list of attempted removals of keys which map to C which is checked before decrementing the counter, as if we had only the counter someone could incorrectly remove A twice and then get a false negative on B.